sstables/storage.cc

```diff
@@ -114,6 +117,43 @@ future<> filesystem_storage::rename_new_file(const sstable& sst, sstring from_na
     });
 }
 
+future<> filesystem_storage::sync_cached_directory(std::string_view name) const {
+    std::string key(name);
```

Comment:
hmm, another nit picked... sorry.
I think that `key` is not really needed.
`name` can be passed to `find` and to `_cached_dirs.emplace` for constructing the std::string key in-place in the unordered_map itself.
And the warning can accept `name` as well, of course.
Just call the parameter `key` :)
And don't move it to `emplace`...

Otherwise, this looks fine.

While at it, if we're already splitting hairs.
If we call `co_await dir_f.flush()` before the emplacement into _cached_dirs, we would even attempt to cache it if flush failed, and the file will be closed when dir_f is destroyed (as it would happen when _cached_dirs is destroyed by the way.  and we may want to destroy it more gently if it might contain a large number of open directories, even using `utils::clear_gently`, see `utils/stall_free.hh`)
Anyhow, if dir_f is flushed before insertion, all we need to do after the failure, is co_await dir_f.close() (out of courtesy), and the co_return unconditionally, and then we don't even need to assign `cached_dir = inserted_at`.

The reason I even thought about that was that I wasn't sure if this assignment might throw or not (it doesn't).

Reply:
> hmm, another nit picked... sorry. I think that `key` is not really needed. `name` can be passed to `find` and to `_cached_dirs.emplace` for constructing the std::string key in-place in the unordered_map itself. And the warning can accept `name` as well, of course. Just call the parameter `key` :) And don't move it to `emplace`...

- If `find()` succeeds, then the number of `std::string` objects convert-constructed from the `std::string_view` parameter `name` is 1, regardless of whether I convert-construct an explicit `std::string` local object `key`, or allow a temporary to be convert-constructed for the argument to `find()`.
- If `find()` fails however, then for the insertion into the unordered map, we need a permanent (long-term) `std::string` convert-constructed from the `std::string_view` parameter `name`. To my understanding, your approach would create a separate copy (yes, it would be constructed in-place, but the `std::string_view` parameter would have to be converted separately again). Whereas as-written, for the insertion, we move-construct the string that's going to be saved in the map, by moving the internals of `key`.

I.e., if `find()` succeeds, then both approaches take the same number of copies (1), but if `find()` fails -- and insertion becomes necessary --, then the code as-written still only takes 1 copy, but your proposal (AIUI) would take 2 (one for the lookup, another for insertion). Avoiding that second copy for insertion is the reason I proposed the code like this.

If I'm wrong about this (or just the style is too complicated), please confirm, and I'll update the code.

> 
> Otherwise, this looks fine.
> 
> While at it, if we're already splitting hairs. If we call `co_await dir_f.flush()` before the emplacement into `_cached_dirs`, we wouldn't even attempt to cache it if flush failed,

I agree that this is possible to do. My intent was that the function should express "flush cached directory", regardless of whether the directory being flushed was added to the cache earlier, or just now. Hence the common call at the end of the function, going through the iterator.

Do you want me to rearrange `sync_cached_directory` like this?

> and the file will be closed when dir_f is destroyed

The pre-patch code contains a similarly explicit (redundant?) close call, going back to commit 8f417b8 ("sstable: coroutinize seal_sstable", 2022-02-14).

Furthermore, regarding exceptions: if we had a checked file impl underneath, and close failed as part of the destructor, wouldn't that cause us to throw from the destructor?

> (as it would happen when _cached_dirs is destroyed by the way. and we may want to destroy it more gently if it might contain a large number of open directories, even using `utils::clear_gently`, see `utils/stall_free.hh`)

IIUC, `clear_gently`, on a `MapLike` container, iterates over the container with `do_until`, and erases elements one by one, inserting a yield point after each element.

Where would we call `clear_gently` on the directory cache? The `sstables::storage` base class from `sstables/storage.hh` has an explicitly defined, empty-body virtual destructor; the derived class `sstables::filesystem_storage` inherits that (empty) destructor, but does not override it. Should we introduce a specialized destructor for `sstables::filesystem_storage`, for calling `clear_gently`?

In turn: is it permitted to use `co_await` in destructors? (And if so, if an individual file close action throws, what's the idiomatic way to suppress that in a destructor?)

> Anyhow, if dir_f is flushed before insertion, all we need to do after the failure, is co_await dir_f.close() (out of courtesy), and co_return unconditionally, also on success, and then we don't even need to assign `cached_dir = inserted_at`.

Right. As I stated, this aspect of the function was not accidental; it was my intent to express a shared "flush *cached* directory" code path at the end (i.e., the dir being flushed must exist in the cache one way or another). If you think that's not a good idea, I'm happy to rewrite the function.

> The reason I even thought about that was that I wasn't sure if this assignment might throw or not (it doesn't).

Right, iterator assignment should be (nearly) trivial.

Reply:
> If I'm wrong about this (or just the style is too complicated), please confirm, and I'll update the code.

I don't see where the second copy is needed?
`auto cached_dir = _cached_dirs.find(key);` uses the string_view
`_cached_dirs.emplace(std::move(key)` can copy the string_view into the map key
`sstlog.warn("Failed to insert {} into cache: {}", key, ex);` shouldn't copy `key` either as for printing its contents, the string_view is just fine.

Reply:
> Where would we call `clear_gently` on the directory cache? The `sstables::storage` base class from `sstables/storage.hh` has an explicitly defined, empty-body virtual destructor; the derived class `sstables::filesystem_storage` inherits that (empty) destructor, but does not override it. Should we introduce a specialized destructor for `sstables::filesystem_storage`, for calling `clear_gently`?

Let's consider that as a possible follow up.
We should be probably add `virtual future<> stop()` to `sstables::storage` that could do the gentle cleanup work.  But let's not hold up this PR on that...
Cc @xemul

Reply:
> In turn: is it permitted to use `co_await` in destructors? (And if so, if an individual file close action throws, what's the idiomatic way to suppress that in a destructor?)

No, it's not.  We'd have to do that in a background thread.
See a bad historical example in:
```c++
vnode_effective_replication_map::~vnode_effective_replication_map() {
    if (is_registered()) {
        _factory->erase_effective_replication_map(this);
        try {
            _factory->submit_background_work(clear_gently(std::move(_replication_map),
                std::move(_pending_endpoints),
                std::move(_read_endpoints),
                std::move(_tmptr)));
        } catch (...) {
            // ignore
        }
    }
}
```

Reply:
@bhalevy 

> I don't see where the second copy is needed? `auto cached_dir = _cached_dirs.find(key);` uses the string_view

I disagree; `std::unordered_map::find()` takes a `const Key& key`, and `Key` is `std::string` in this case. So, for the lookup's sake, an `std::string` needs to be constructed from the `std::string_view` input parameter.

> `_cached_dirs.emplace(std::move(key)` can copy the string_view into the map key

That would only work if the `Key` type of the map were `std::string_view` -- but `std::string_view` is non-owning, so I'm not convinced it's safe to store in the map.

Reply:
> Let's consider that as a possible follow up.
We should be probably add virtual future<> stop() to sstables::storage that could do the gentle cleanup work. But let's not hold up this PR on that...
Cc @xemul

There's `storage::destroy()` for that. The storage object dies with an sstable and this method is called prior to that

Reply:
resolving as obsolete per https://github.com/scylladb/scylladb/pull/19624#issuecomment-2218322451